### PR TITLE
JDK-8301493: Replace NULL with nullptr in cpu/aarch64

### PR DESCRIPTION
**Pull request by jdksjolen**
Hi, this PR changes all occurrences of NULL to nullptr for the subdirectory cpu/aarch64. Unfortunately the script that does the change isn't perfect, and so we
need to comb through these manually to make sure nothing has gone wrong. I also review these changes but things slip past my eyes sometimes.

Here are some typical things to look out for:

1. No changes but copyright header changed (probably because I reverted some changes but forgot the copyright).
2. Macros having their NULL changed to nullptr, these are added to the script when I find them. They should be NULL.
3. nullptr in comments and logs. We try to use lower case "null" in these cases as it reads better. An exception is made when code expressions are in a comment.

An example of this:

```c++
// This function returns null
void* ret_null();
// This function returns true if *x == nullptr
bool is_nullptr(void** x);
```

Note how `nullptr` participates in a code expression here, we really are talking about the specific value `nullptr`.

Thanks!

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301493](https://bugs.openjdk.org/browse/JDK-8301493): Replace NULL with nullptr in cpu/aarch64

### Reviewers
 * [Thomas Schatzl](https://openjdk.org/census#tschatzl) (@tschatzl - **Reviewer**)
 * [Gerard Ziemski](https://openjdk.org/census#gziemski) (@gerard-ziemski - Committer) ⚠️ Review applies to [5cc7a5a7](https://git.openjdk.org/jdk/pull/12321/files/5cc7a5a7641132aab36bb24ac7c2701a89f7eff8)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [5cc7a5a7](https://git.openjdk.org/jdk/pull/12321/files/5cc7a5a7641132aab36bb24ac7c2701a89f7eff8)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/12321/head:pull/12321` \
`$ git checkout pull/12321`

Update a local copy of the PR: \
`$ git checkout pull/12321` \
`$ git pull https://git.openjdk.org/jdk.git pull/12321/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12321`

View PR using the GUI difftool: \
`$ git pr show -t 12321`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12321.diff">https://git.openjdk.org/jdk/pull/12321.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/12321#issuecomment-1418846528)

### CODE REVIEW

File: src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp [hunk 1]
```diff
@@ -171,13 +171,13 @@ inline frame::frame(intptr_t* sp, intptr_t* fp) {
   // call a specilaized frame constructor instead of this one.
   // Then we could use the assert below. However this assert is of somewhat dubious
   // value.
-  // assert(_pc != NULL, "no pc?");
+  // assert(_pc != null, "no pc?");
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/frame_aarch64.inline.hpp:174]: nullptr

File: src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp [hunk 2]
```diff
@@ -159,7 +159,7 @@ void G1BarrierSetAssembler::g1_write_barrier_pre(MacroAssembler* masm,
 
   // Calling the runtime using the regular call_VM_leaf mechanism generates
   // code (generated by InterpreterMacroAssember::call_VM_leaf_base)
-  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == NULL.
+  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == null.
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/gc/g1/g1BarrierSetAssembler_aarch64.cpp:162]: nullptr

File: src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp [hunk 3]
```diff
@@ -157,7 +157,7 @@ void ShenandoahBarrierSetAssembler::satb_write_barrier_pre(MacroAssembler* masm,
 
   // Calling the runtime using the regular call_VM_leaf mechanism generates
   // code (generated by InterpreterMacroAssember::call_VM_leaf_base)
-  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == NULL.
+  // that checks that the *(rfp+frame::interpreter_frame_last_sp) == null.
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/gc/shenandoah/shenandoahBarrierSetAssembler_aarch64.cpp:160]: nullptr

File: src/hotspot/cpu/aarch64/icBuffer_aarch64.cpp [hunk 4]
```diff
@@ -47,7 +47,7 @@ void InlineCacheBuffer::assemble_ic_buffer_code(address code_begin, void* cached
   // because
   // (1) the value is old (i.e., doesn't matter for scavenges)
   // (2) these ICStubs are removed *before* a GC happens, so the roots disappear
-  // assert(cached_value == NULL || cached_oop->is_perm(), "must be perm oop");
+  // assert(cached_value == null || cached_oop->is_perm(), "must be perm oop");
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/icBuffer_aarch64.cpp:50]: nullptr

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 5]
```diff
@@ -140,7 +140,7 @@ void InterpreterMacroAssembler::check_and_handle_earlyret(Register java_thread)
   if (JvmtiExport::can_force_early_return()) {
     Label L;
     ldr(rscratch1, Address(rthread, JavaThread::jvmti_thread_state_offset()));
-    cbz(rscratch1, L); // if (thread->jvmti_thread_state() == NULL) exit;
+    cbz(rscratch1, L); // if (thread->jvmti_thread_state() == null) exit;
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:143]: nullptr

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 6]
```diff
@@ -1303,13 +1303,13 @@ void InterpreterMacroAssembler::record_item_in_profile_helper(Register item, Reg
...
+//   if (row[0].rec != null) {
 //     // inner copy of decision tree, rooted at row[1]
 //     if (row[1].rec == rec) { row[1].incr(); goto done; }
-//     if (row[1].rec != NULL) {
+//     if (row[1].rec != null) {
 //       // degenerate decision tree, rooted at row[2]
 //       if (row[2].rec == rec) { row[2].incr(); goto done; }
-//       if (row[2].rec != NULL) { count.incr(); goto done; } // overflow
+//       if (row[2].rec != null) { count.incr(); goto done; } // overflow
 //       row[2].init(rec); goto done;
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:1313]: nullptr

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 7]
```diff
@@ -1583,7 +1583,7 @@ void InterpreterMacroAssembler::call_VM_leaf_base(address entry_point,
     ldr(rscratch1, Address(rfp, frame::interpreter_frame_last_sp_offset * wordSize));
     cbz(rscratch1, L);
     stop("InterpreterMacroAssembler::call_VM_leaf_base:"
-         " last_sp != NULL");
+         " last_sp != null");
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:1586]: nullptr

File: src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp [hunk 8]
```diff
@@ -1611,7 +1611,7 @@ void InterpreterMacroAssembler::call_VM_base(Register oop_result,
     ldr(rscratch1, Address(rfp, frame::interpreter_frame_last_sp_offset * wordSize));
     cbz(rscratch1, L);
     stop("InterpreterMacroAssembler::call_VM_base:"
-         " last_sp != NULL");
+         " last_sp != null");
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/interp_masm_aarch64.cpp:1614]: nullptr

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 9]
```diff
@@ -1155,7 +1155,7 @@ void MacroAssembler::lookup_interface_method(Register recv_klass,
       add(recv_klass, recv_klass, itentry_off);
   }
 
-  // for (scan = klass->itable(); scan->interface() != NULL; scan += scan_step) {
+  // for (scan = klass->itable(); scan->interface() != null; scan += scan_step) {
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1158]: nullptr

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 10]
```diff
@@ -4788,15 +4788,15 @@ address MacroAssembler::arrays_equals(Register a1, Register a2, Register tmp3,
   br(EQ, SAME);
 
   if (UseSimpleArrayEquals) {
-    Label NEXT_WORD, SHORT, TAIL03, TAIL01, A_MIGHT_BE_NULL, A_IS_NOT_NULL;
+    Label NEXT_WORD, SHORT, TAIL03, TAIL01, A_MIGHT_BE_nullptr, A_IS_NOT_NULL;
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:4791]: Fix this macro

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 11]
```diff
@@ -2044,19 +2044,19 @@ class StubGenerator: public StubCodeGenerator {
...
 
-    //  if (src == NULL) return -1;
+    //  if (src == null) return -1;
     __ cbz(src, L_failed);
 
     //  if (src_pos < 0) return -1;
     __ tbnz(src_pos, 31, L_failed);  // i.e. sign bit set
 
-    //  if (dst == NULL) return -1;
+    //  if (dst == null) return -1;
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:2059]: nullptr

File: src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp [hunk 12]
```diff
@@ -2073,11 +2073,11 @@ class StubGenerator: public StubCodeGenerator {
 
     __ load_klass(scratch_src_klass, src);
 #ifdef ASSERT
-    //  assert(src->klass() != NULL);
+    //  assert(src->klass() != null);
     {
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp:2077]: nullptr

File: src/hotspot/cpu/aarch64/templateTable_aarch64.cpp [hunk 13]
```diff
@@ -3694,8 +3694,8 @@ void TemplateTable::instanceof() {
     __ bind(is_null);   // same as 'done'
   }
   __ bind(done);
-  // r0 = 0: obj == NULL or  obj is not an instanceof the specified klass
-  // r0 = 1: obj != NULL and obj is     an instanceof the specified klass
```
Comment (jdksjolen) [inline src/hotspot/cpu/aarch64/templateTable_aarch64.cpp:3698]: nullptr

**Review by jdksjolen — COMMENTED** [review]
Some simple fixes here

**Comment by jdksjolen** [comment]
Not yet.

**Comment by stooart-mon** [comment]
This looks OK so far. 
However, is it your intention to also do aarch64.ad?
aarch64_ad.m4 and aarch64_vector(.ad|_ad.m4) files look clean.

File: src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp [hunk 14]
```diff
@@ -267,7 +267,7 @@ class SlowSignatureHandler
 
   virtual void pass_object() {
     intptr_t* addr = single_slot_addr();
-    intptr_t value = *addr == 0 ? NULL : (intptr_t)addr;
+    intptr_t value = *addr == 0 ? nullptr : (intptr_t)addr;
```
Comment (stooart-mon) [inline src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp:270]: This doesn't compile - perhaps replace nullptr with zero? Unless casting it is more appropriate.

Comment (jdksjolen) [reply src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp]: There's similar code in other patches, I did casting in those.

**Review by tschatzl — CHANGES_REQUESTED** [review]
Remaining `NULL` in

gc/shared/BarrierSetAssembler::check_oop()
codeBuffer_aarch64.cpp/emit_shared_trampolines()
stubGenerator_aarch64.cpp/generate_final_stubs()
vm_version_aarch64.cpp/check_info_file()

File: src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp [hunk 15]
```diff
@@ -267,7 +267,7 @@ class SlowSignatureHandler
 
   virtual void pass_object() {
     intptr_t* addr = single_slot_addr();
-    intptr_t value = *addr == 0 ? NULL : (intptr_t)addr;
+    intptr_t value = *addr == 0 ? (intptr_t)nullptr : (intptr_t)addr;
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/interpreterRT_aarch64.cpp:270]: This looks like it should be using 0 (zero) not NULL/nullptr?

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 16]
```diff
@@ -968,7 +968,7 @@ void MacroAssembler::emit_static_call_stub() {
   // exact layout of this stub.
 
   isb();
-  mov_metadata(rmethod, (Metadata*)NULL);
+  mov_metadata(rmethod, (Metadata*)nullptr);
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:971]: Shouldn't need cast any more.

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 17]
```diff
@@ -1639,13 +1639,13 @@ void MacroAssembler::super_call_VM_leaf(address entry_point, Register arg_0, Reg
 
 void MacroAssembler::null_check(Register reg, int offset) {
   if (needs_explicit_null_check(offset)) {
-    // provoke OS NULL exception if reg = NULL by
+    // provoke OS null exception if reg = null by
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1642]: Suggest `reg is null`

File: src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp [hunk 18]
```diff
@@ -1639,13 +1639,13 @@ void MacroAssembler::super_call_VM_leaf(address entry_point, Register arg_0, Reg
...
     // accessing M[reg] w/o changing any registers
     // NOTE: this is plenty to provoke a segv
     ldr(zr, Address(reg));
   } else {
     // nothing to do, (later) access of M[reg + offset]
-    // will provoke OS NULL exception if reg = NULL
+    // will provoke OS null exception if reg = null
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp:1648]: Suggest `reg is null`

File: src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp [hunk 19]
```diff
@@ -1421,10 +1421,10 @@ nmethod* SharedRuntime::generate_native_wrapper(MacroAssembler* masm,
                                        stack_slots / VMRegImpl::slots_per_word,
                                        in_ByteSize(-1),
                                        in_ByteSize(-1),
-                                       (OopMapSet*)NULL);
+                                       (OopMapSet*)nullptr);
```
Comment (dholmes-ora) [inline src/hotspot/cpu/aarch64/sharedRuntime_aarch64.cpp:1424]: Shouldn't need cast any more

**Review by dholmes-ora — APPROVED** [review]
Looks good - thanks!

Two minor suggested changes and four opportunities to remove casts (that I spotted).

**Comment by jdksjolen** [comment]
Thanks @tschatzl, probably new code from when I merged that didn't get a conflict. I did a grep for NULL and checked that no more occurrences are in the code.  Running tier1 testing.

**Review by tschatzl — APPROVED** [review]
Lgtm.

**Comment by jdksjolen** [comment]
Alright, tier1 looks good. Integrating. Thank you for the reviews, good people.

/integrate

**Pull request closed without merging**